### PR TITLE
enable shared private and public maps

### PR DESCRIPTION
**Pull request by Connoropolous**
conversation originally iin #424 

- [x] update Policy scopes
- [x] update 'change map permission' text in the Map Info Box
- [x] adjust JS so that topics and synapses can be edited if their 'defer_to_map' can be
- [x] when a topic or synapse has 'defer_to_map' and gets removed from that map, remove 'defer_to_map' and set to latest simple permission of the map

### CODE REVIEW

**Comment by Connoropolous** [comment]
- [ ] enable realtime, chat and Junto on private and public maps

**Comment by devvmh** [comment]
:open_mouth:

**Comment by Connoropolous** [comment]
- [ ] update Policy scopes

**Comment by Connoropolous** [comment]
- [ ] handle Junto panel whenever you don't have 'write to map' access

File: app/policies/map_policy.rb [hunk 1]
```diff
@@ -4,7 +4,8 @@ def resolve
       visible = ['public', 'commons']
       permission = 'maps.permission IN (?)'
       if user
-        scope.where(permission + ' OR maps.user_id = ?', visible, user.id)
+        shared_maps = user.user_maps.map(&:id)
```
Comment (Connoropolous) [inline app/policies/map_policy.rb]: @devvmh this is where the magic happens

File: app/policies/synapse_policy.rb [hunk 2]
```diff
@@ -4,7 +4,7 @@ def resolve
       visible = ['public', 'commons']
       permission = 'synapses.permission IN (?)'
       if user
-        scope.where(permission + ' OR synapses.user_id = ?', visible, user.id)
+        scope.where(permission + ' OR synapses.defer_to_map_id IN (?) OR synapses.user_id = ?', visible, user.shared_maps.map(&:id), user.id)
```
Comment (Connoropolous) [inline app/policies/synapse_policy.rb]: @devvmh and here

**Comment by Connoropolous** [comment]
guh, this'll be a fun one to write tests for :grimacing:

**Comment by Connoropolous** [comment]
taken 6 hours so far, so not bad. it's mostly done. 

realistically it would have been horrible for anyone besides me to write this... have to try and improve that... prbly with tests :p

File: app/controllers/topics_controller.rb [hunk 3]
```diff
@@ -150,7 +150,7 @@ def destroy
     @topic = Topic.find(params[:id])
     authorize @topic
 
-    @topic.delete
+    @topic.destroy
```
Comment (Connoropolous) [inline app/controllers/topics_controller.rb]: @devvmh this is interesting: I discovered that by using `delete` here, our `dependent: :destroy` methods weren't taking effect and deleting mappings, so this is actually a bug on prod... we've got orphaned data happening

**Comment by devvmh** [comment]
@Connoropolous happily, it looks like no one has deleted any topics/synapses since the rails 4 update:

````
2.1.3 :002 > Mapping.where(mappable_id: nil).count
 => 0
2.1.3 :003 > Mapping.where(mappable_type: nil).count
 => 0
````

**Comment by Connoropolous** [comment]
screen shot 2016-04-13 at 8 33 51 pm

**Comment by devvmh** [comment]
Hey @Connoropolous , check out https://github.com/metamaps/metamaps/tree/feature/merge-with-connor

This merges develop and then it merges my branch feature/refactor-javascript.

If you do `git reset --hard feature/merge-with-connor`, I think it'll work the same and be rebased on that branch, so then i could merge #548.

But I'm not 100% sure that it's all perfect, so I'd want you to test it a bit before going ahead

**Comment by Connoropolous** [comment]
Kk, true @devvmh I will check them both out and try and get them in

File: app/policies/map_policy.rb (hunk 1 above)
Comment (Connoropolous) [inline app/policies/map_policy.rb]: Note to self &:id should be &:map_id in actuality I believe... Currently its a bug

**Comment by Connoropolous** [comment]
Thanks Devin! Helps for making it so much easier to review. 
And yes, your last commit does fix the bug I mentioned!

Just reporting one more here, for my own sake, in the addTypeahead js function within Metamaps.Map.js

**Comment by Connoropolous** [comment]
@devvmh I've improved the 'unauthorized' response slightly in this branch

**Pull request merged**